Use Rollup and ES modules (v3)

donmccurdy opened the pull request:
Re-opening #116. Includes `dist/*.js` files.

donmccurdy commented:
Copying over a few notes from #116 —

Future PRs (e.g. by community members) should omit the dist/ files, because they're hard to review and lead to merge conflicts — it's easier just to update them when needed for a new version or demo deployment.

Do note that with this change, NPM (or an alternative like https://yarnpkg.com/) will be part of the development process, and should be running in the background (npm run dev) while working locally. Static hosting for the demo is still supported of course, NPM does not need to be installed on the server. It's enough just to run npm run build and then push the files to a location like GitHub Pages.

Feel free to leave this open while reviewing or discussing. Once this is checked in I'm also happy to add a hook so that future PRs will display a warning if they break the build script.

emackey commented:
**NOTE** Because this Pull Request changes the typical daily workflow for developers on this project, it would be very helpful if everyone working on this reference viewer could give this branch a quick test and make sure that development can continue under this new system.

Specifically:

* All developers on this project should have NodeJS installed (@donmccurdy any particular version?  I usually recommend the latest LTS)
* Run `npm install` to bring down the development dependencies.
* Run `npm run dev` to launch the development server, so that changes to the source code can be picked up interactively.
* As needed, run `npm run build` when the `dist/` folder needs to be re-compiled, for incorporating changes into a static-file deployment.  Typically this would be part of a release process, but not included with individual future PRs, for reasons stated above.

Please leave feedback here with any comments or questions.

donmccurdy commented:
> All developers on this project should have NodeJS installed.

I authored the PR with v8.11.4, but would expect anything >6.x, and certainly current LTS, to work correctly.

UX3D-schmithuesen commented:
Is there a reason to have the dist files in the repository then or can we remove and gitignore them?

emackey commented:
@UX3D-schmithuesen The final project is hosted via GitHub Pages, so the `dist/` files need to be compiled and included on the GitHub branch that is being served (which is currently `master`).  Don mentioned that pull requests should not check in any changes to `dist`, so as not to produce merge conflicts.

@donmccurdy Thinking through this a bit further, it seems like the Pages deployment could get out-of-sync, if PRs start coming into master without `dist` updates.  Much earlier in the life of this project, we chose to deploy from `master` instead of `gh-pages` just for the convenience of automatically deploying anything that got merged.  Could we add a hook to automatically recompile `dist` after any merge?  Or maybe we should reconfigure to serve from `gh-pages` again, and ignore `dist` on master but not on gh-pages, so that deployments can happen by merging master to gh-pages and compiling.  The latter approach has the advantage that developers could compile locally without producing Git changes.

donmccurdy commented:
> Is there a reason to have the dist files in the repository then or can we remove and gitignore them?

I'd say it's more common to commit them in JS projects, but removing and ignoring them would be fine. The downsides are (1) we'll need to adjust the Pages deployment, which shouldn't be an issue, and (2) users occasionally request to be able to download a single script from GitHub without compiling the project themselves.

> it seems like the Pages deployment could get out-of-sync, if PRs start coming into master without dist updates...

That depends on what we want the Pages demo to reflect — current master branch, or the latest stable release? I think either is fine.

> Could we add a hook to automatically recompile dist after any merge? Or maybe we should reconfigure to serve from gh-pages again, and ignore dist on master but not on gh-pages, so that deployments can happen by merging master to gh-pages and compiling. 

Yes, we can do either of these. Reconfiguring to use `gh-pages` again would keep the Git history cleaner, so that might be my preference. I'm not sure whether this automation (rebuild --> test --> rebuild --> deploy) can be done as easily on the `reference-viewer` branch, or would need to wait until we've moved development to `master`.

UX3D-schmithuesen commented:
Apart from the git history, I also like to keep my working directory clean and find it strange to always have 3 modified files that I shouldn't commit.
@donmccurdy can you check if the automation is possible on the `reference-viewer` branch? That sounds like the best option to me.

donmccurdy commented:
Shouldn't be any problem there, let's use CircleCI (as in glTF-Blender-IO) and configure it per https://blog.frederikring.com/articles/deploying-github-pages-circle-ci/.

Using `gh-pages` will overwrite http://github.khronos.org/glTF-WebGL-PBR/ — is that OK? I think so, but just checking.

I'll remove the `dist/` files from this PR and add them to `.gitignore`, then?

emackey commented:
> Using `gh-pages` will overwrite http://github.khronos.org/glTF-WebGL-PBR/ — is that OK? I think so, but just checking.

We will need James to make this change (or rather, to undo it).  Currently the repository is configured to serve GitHub Pages from the `master` branch, not the `gh-pages` branch.  James can fix this for us if we ask him to do so.

> I'll remove the `dist/` files from this PR and add them to `.gitignore`, then?

Yeah, I'm thinking that's better.

donmccurdy commented:
I've removed the `dist/` files and added a `.npmignore` file. The latter is required to tell NPM to include files that Git does not.

Until we're ready to use GitHub Pages, I don't know how to set up automation pushing `reference-viewer` to the http://gltf.ux3d.io/ URL. Someone else may have to do that, or keep deploying manually if that's what's happening now.

UX3D-schmithuesen commented:
So is this ready to be merged? Seems fine to me!

donmccurdy commented:
Yep, I believe this PR is ready.
